java/source/jni/dataset.rst

```diff
@@ -0,0 +1,294 @@
+
+    import java.util.List;
+    import java.util.Optional;
+    import java.util.stream.Collectors;
+    import java.util.stream.StreamSupport;
+
+    String uri = "file:" + System.getProperty("user.dir") + "/thirdpartydeps/parquetfiles/data1.parquet";
+    RootAllocator rootAllocator = new RootAllocator(Long.MAX_VALUE);
+    DatasetFactory datasetFactory = new FileSystemDatasetFactory(rootAllocator, NativeMemoryPool.getDefault(), FileFormat.PARQUET, uri);
+    String[] projection = new String[] {"name"};
+    ScanOptions options = new ScanOptions(1, Optional.of(projection));
+    Dataset dataset = datasetFactory.finish();
+    Scanner scanner = dataset.newScan(options);
+    Schema schema = scanner.schema();
+    List<ArrowRecordBatch> batches = StreamSupport.stream(scanner.scan().spliterator(), false).flatMap(t -> Streams.stream(t.execute())).collect(Collectors.toList());
+    int fieldCount = schema.getFields().size();
+    try (VectorSchemaRoot vsr = VectorSchemaRoot.create(schema, rootAllocator)) {
+        VectorLoader loader = new VectorLoader(vsr);
+        for (ArrowRecordBatch batch : batches) {
+            loader.load(batch);
+            int batchRowCount = vsr.getRowCount();
+            for (int i = 0; i < fieldCount; i++) {
+                FieldVector vector = vsr.getVector(i);
+                for (int j = 0; j < batchRowCount; j++) {
+                    Object object = vector.getObject(j);
+                    System.out.println(object);
+                }
+            }
+        }
+    }
+    AutoCloseables.close(datasetFactory, dataset, scanner);
+
+.. testoutput::
+
+    David
+    Gladis
+    Juan
+
+JNI Library
+===========
+
+Shared Libraries / Dynamically Loaded Modules
+*********************************************
+
+Consider arrow dataset versions for jni libarrow_dataset_jni.dylib and libarrow_dataset_jni.so.
+
+.. code-block:: shell
+
+    (base) ➜  /tmp jar -tf ~/Downloads/arrow-dataset-6.0.0.jar|grep _jni.
+
+    libarrow_dataset_jni.dylib
+    libarrow_dataset_jni.600.0.0.dylib
+    libarrow_dataset_jni.600.dylib
+
+    (base) ➜  /tmp jar -tf ~/Downloads/arrow-dataset-6.0.1.jar|grep _jni.
+
+    libarrow_dataset_jni.so
+    libarrow_dataset_jni.so.600
+    libarrow_dataset_jni.so.600.1.0
+
+Library Dependencies
+********************
+
+Consider this for jni libarrow_dataset_jni.dylib and libarrow_dataset_jni.so library dependencies.
+
+Shared Libraries:
+
+.. code-block:: shell
+
+    (base) ➜  /tmp objdump -p libarrow_dataset_jni.so | grep NEEDED
+      NEEDED       liblz4.so.1
+      NEEDED       libsnappy.so.1
+      NEEDED       libz.so.1
+      NEEDED       libzstd.so.1
+      NEEDED       libutf8proc.so.2
+      NEEDED       libre2.so.9
+      NEEDED       libthrift-0.13.0.so
+      NEEDED       libstdc++.so.6
+      NEEDED       libm.so.6
+      NEEDED       libgcc_s.so.1
+      NEEDED       libpthread.so.0
+      NEEDED       libc.so.6
+      NEEDED       ld-linux-x86-64.so.2
```

Comment:
To be frank, I would have expected JARs with binaries to be self-contained, so this is a little unfortunate. Maybe we need to resolve this on the packaging side?

Reply:
If we offer fat jar with all dependency included the problem could be the same if the user does not have that dependencies already configured/installed.

For example if the fat jar contains libc.so.6 and the user OS is ubuntu 20 it wont work because the  libc.so.6 needs ubuntu 21 that contains GLIBC_2.32, if not the user could obtain error message like this `./libarrow_dataset_jni.so: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by ./libarrow_dataset_jni.so)`

Reply:
Well then, it's mostly concerning that we're building JARs that sound practically useless (especially as it is not obvious what dependencies/system requirements exactly they have), and it would be much better if we could keep the cookbook organization instead of having to split things like this. But if testing on MacOS is the only way, then so be it.

Could we just run a single pipeline on MacOS, or at least run all the non-JNI tests on both platforms so we're not doing as much work in the CI to try to split up the tests so carefully?

Reply:
We are able to run jni dataset 7.0.0 version on latest github ci images without problems and not needed to install or upgrade custom dependencies.